### DISCUSSION TITLE
Uploading an attachment to a BudibaseDB row using the API

### DISCUSSION THREAD

**Discussion by daawaan4x** [Budibase/budibase#4995] [Help] [answered]
We've been tasked to migrate 3000+ pdf files which consist scanned hard-copies of documents into their corresponding rows in the BudibaseDB. The API references in the docs describe how to update a field in a row by a following a ```field: value``` format, but what would be the format for updating an attachment column?

**Comment by Samirmabrouk** [comment]
Hämta Outlook för iOS<https://aka.ms/o0ukef>
________________________________
Från: Theone Eclarin ***@***.***>
Skickat: Tuesday, March 22, 2022 4:00:35 AM
Till: Budibase/budibase ***@***.***>
Kopia: Subscribed ***@***.***>
Ämne: [Budibase/budibase] Uploading an attachment to a BudibaseDB row using the API (Discussion #4995)


We've been tasked to migrate 3000+ pdf files which consist scanned hard-copies of documents into their corresponding rows in the BudibaseDB. The API references in the docs describe how to update a field in a row by a following a field: value format, but what would be the format for updating an attachment column?

**Reply by mjashanks** [reply]
@Samirmabrouk is this some sort of automated email reply here? It's spammy - could you please turn this off?

**Comment by mjashanks** [comment]
Hey @daawaan, the best way to find out how to do this is to:

1. Create form (in Budibase), that has an attachment field
2. Open your app
3. Open your dev tools > network tab
4. Upload an attachment and save the record.

By watching the network requests, you should be able to figure out how to do this. The process is roughly
- Upload attachment to server - attachment URL / metadata is returned
- Save attachment metadata to the Attachment field

**Comment by xnohat** [comment, marked as answer]
Hi,
Below is complete script I use for upload attachment through API

```
const fs = require('fs');
const fetch = require('node-fetch'); //node-fetch@2.6.1
const FormData = require('form-data');

/*
args:
0 : bbDomain ex: budibase.app
1 : apikey ex: 473fc3f164a9662f8c27402375ca3e26-e200caba55ec73e0d0cb00347b2045462e13cda7a47614e8888e80ac1b505c8801d5c1b0e02c2586
2 : bbAppId ex: app_2d2b460e23f7400b90a94e94190d6ff2
3 : bbRowId ex: ro_ta_0c307ee1d8e64e8481e7202a9ba84984_6a8d9c6878a141a39a8babd55d2d0890
4 : bbColumnName ex: "Attachment File"
5 : path/to/file-need-upload.ext

Example: node uploadtobudi.js budibase.app 473fc3f164a9662f8c27402375ca3e26-e200caba55ec73e0d0cb00347b2045462e13cda7a47614e8888e80ac1b505c8801d5c1b0e02c2586 app_2d2b460e23f7400b90a94e94190d6ff2 ro_ta_0c307ee1d8e64e8481e7202a9ba84984_b4d6764a8d91483a98b02bdd7a0f9bb3 "Audio File" thereaderlogo.jpg
*/
var args = process.argv.slice(2);

//budibase api conf
const bbdomain = args[0];
const bbapikey = args[1];
const bbappid = args[2]; //make use using Prod App ID with data from Prod or you must re-Publish app first
const bbrowid = args[3];
const bbtableid = bbrowid.match(/(ta_.+)_/)[1]; //extract from bbrowid
const bbcolumn = args[4];

//Upload File
const filePath = args[5];
const form = new FormData();
const stats = fs.statSync(filePath);
const fileSizeInBytes = stats.size;
const fileStream = fs.createReadStream(filePath);
form.append('file', fileStream, { knownLength: fileSizeInBytes });

async function uploadtoBudi() {

const options_for_upload = {
    "method": "POST",
    "headers": {
        "x-budibase-app-id": bbappid,
        "x-budibase-api-key": bbapikey,
    },
    "body": form,
    };

var uploadedfile = await (await fetch(`https://${bbdomain}/api/attachments/${bbtableid}/upload`, { ...options_for_upload })).json();

//Update Row with file uploaded
var new_row_data_for_update = `{
    "_id": "${bbrowid}",
    "tableId": "${bbtableid}",
    "type": "row",
    "${bbcolumn}": ${JSON.stringify(uploadedfile)}
}`.replace(/(\r\n|\n|\r)/gm, "").replace(/("(?:\\"|[^"])*")|\s/g, "$1");

const options_for_update_row = {
    "method": "PUT",
    "headers": {
        "Accept": 'application/json',
        "Content-Type": 'application/json',
        "x-budibase-app-id": bbappid,
        "x-budibase-api-key": bbapikey
    },
    "body": new_row_data_for_update,
  };

var updatedrow = await (await fetch(`https://${bbdomain}/api/public/v1/tables/${bbtableid}/rows/${bbrowid}`, { ...options_for_update_row })).json();

return updatedrow;

}

//Execute
uploadtoBudi()
    .then(data => console.log(data))
    .catch(reason => console.log(reason.message));
```

**Reply by mjashanks** [reply]
Thank you for sharing!
